test/cpp_api_parity/sample_module.py

```diff
@@ -0,0 +1,84 @@
+import torch
+from cpp_api_parity import CppArgDeclaration
+
+'''
+`SampleModule` is used by `test_cpp_api_parity.py` to test that Python / C++ API
+parity test harness works for `torch.nn.Module` subclasses.
+
+When `SampleModule.has_parity` is true, behavior of `reset_parameters` / `forward` /
+`backward` is the same as the C++ equivalent.
+
+When `SampleModule.has_parity` is false, behavior of `reset_parameters` / `forward` /
+`backward` is different from the C++ equivalent.
+'''
+
+class SampleModule(torch.nn.Module):
+    def __init__(self, has_parity):
+        super(SampleModule, self).__init__()
+        self.register_parameter('param', torch.nn.Parameter(torch.empty(3, 4)))
+        self.register_buffer('buffer', torch.empty(4, 5))
+        self.has_parity = has_parity
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        with torch.no_grad():
+            self.param.fill_(1)
+            self.buffer.fill_(1)
+            if not self.has_parity:
+                self.param.add_(10)
+                self.buffer.add_(10)
+
+    def forward(self, x):
+        if not self.has_parity:
+            return x + self.param * 4 + 3
+        else:
+            return x + self.param * 2
+
+SAMPLE_MODULE_CPP_SOURCE = """\n
+struct SampleModuleImpl : public torch::nn::Cloneable<SampleModuleImpl> {
```

Comment:
(just a thought) How bad would it be if we added a DoNotUseSampleModule into the C++ API and use that for the test?

Right now it's not very nice that a lot of the testing code lives in a C++ extension and not in core PyTorch. If someone were to do something like a codemod or change some parts of the API, it would take them a while to figure out that this file needs to be changed.

Reply:
I feel that we should keep this test code out of the C++ API to keep the API clean. When someone changes some parts of the API that makes the usage here out of date, they will be able to notice it via the cpp extension compile error message when they run this parity test.

Reply:
That's reasonable